### PR TITLE
[SPARK-30083][SQL] visitArithmeticUnary should wrap PLUS case with UnaryPositive for type checking

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

`UnaryPositive` only accepts numeric and interval as we defined, but what we do for this in  `AstBuider.visitArithmeticUnary` is just bypassing it.

This should not be omitted for the type checking requirement.

### Why are the changes needed?

bug fix, you can find a pre-discussion here https://github.com/apache/spark/pull/26578#discussion_r347350398

### Does this PR introduce any user-facing change?
yes,  +non-numeric-or-interval is now invalid.
```
-- !query 14
select +date '1900-01-01'
-- !query 14 schema
struct<DATE '1900-01-01':date>
-- !query 14 output
1900-01-01


-- !query 15
select +timestamp '1900-01-01'
-- !query 15 schema
struct<TIMESTAMP '1900-01-01 00:00:00':timestamp>
-- !query 15 output
1900-01-01 00:00:00


-- !query 16
select +map(1, 2)
-- !query 16 schema
struct<map(1, 2):map<int,int>>
-- !query 16 output
{1:2}


-- !query 17
select +array(1,2)
-- !query 17 schema
struct<array(1, 2):array<int>>
-- !query 17 output
[1,2]


-- !query 18
select -'1'
-- !query 18 schema
struct<(- CAST(1 AS DOUBLE)):double>
-- !query 18 output
-1.0


-- !query 19
select -X'1'
-- !query 19 schema
struct<>
-- !query 19 output
org.apache.spark.sql.AnalysisException
cannot resolve '(- X'01')' due to data type mismatch: argument 1 requires (numeric or interval) type, however, 'X'01'' is of binary type.; line 1 pos 7


-- !query 20
select +X'1'
-- !query 20 schema
struct<X'01':binary>
-- !query 20 output
```

### How was this patch tested?

add ut check

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @HyukjinKwon, thanks for reviewing

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
This makes sense. Can we check the behavior of spark 2.4? If it's a behavior change we need to add a migration guide.

**Comment by yaooqinn** [comment]
In v2.4.4, +(date/timestamp/array/map) are valid, we should add a migration guide for these.
In v2.4.4 and master, +(string) with integrals returns integral, but -(string) with integrals is coerced to double.  we seem to need a migration guide too even through.
bit-string constants with x-prefix is not supported yet.

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

File: docs/sql-migration-guide.md [hunk 1]
```diff
@@ -254,6 +254,8 @@ license: |
         </tr>
     </table>
 
+    - Since Spark 3.0, the unary arithmetic operator plus(`+`) only accept string, numeric and interval type values as inputs. Besides, `+` with a integral string representation will be coerced to double value, e.g. `+'1'` results `1.0`. In Spark version 2.4 and earlier,  there is no type checking for it, thus, all type values with a `+` prefix is valid, e.g. `+ array(1, 2)` is valid and results `[1, 2]`. Besides, there is no implicit type coercion for `+` with string, e.g. `+'1'` results `1`.
```
Comment (maropu) [inline docs/sql-migration-guide.md]: nit: accept -> accepts

**Review by maropu — APPROVED** [review]

File: docs/sql-migration-guide.md [hunk 2]
```diff
@@ -254,6 +254,8 @@ license: |
         </tr>
     </table>
 
+    - Since Spark 3.0, the unary arithmetic operator plus(`+`) only accepts string, numeric and interval type values as inputs. Besides, `+` with a integral string representation will be coerced to double value, e.g. `+'1'` results `1.0`. In Spark version 2.4 and earlier,  there is no type checking for it, thus, all type values with a `+` prefix are valid, e.g. `+ array(1, 2)` is valid and results `[1, 2]`. Besides, there is no implicit type coercion for `+` with string, e.g. `+'1'` results `1`.
+
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: in 2.4, `+'1'` results to string `'1'`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala [hunk 3]
```diff
@@ -226,10 +226,10 @@ class ExpressionParserSuite extends AnalysisTest {
   }
 
   test("unary arithmetic expressions") {
-    assertEqual("+a", 'a)
+    assertEqual("+a", UnaryPositive('a))
     assertEqual("-a", -'a)
     assertEqual("~a", ~'a)
-    assertEqual("-+~~a", -(~(~'a)))
+    assertEqual("-+~~a", -UnaryPositive(~(~'a)))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: shall we create a shortcut '+' for `UnaryPositive` as well? The `-` is defined in `org.apache.spark.sql.catalyst.dsl`

File: docs/sql-migration-guide.md [hunk 4]
```diff
@@ -254,6 +254,8 @@ license: |
         </tr>
     </table>
 
+    - Since Spark 3.0, the unary arithmetic operator plus(`+`) only accepts string, numeric and interval type values as inputs. Besides, `+` with a integral string representation will be coerced to double value, e.g. `+'1'` results `1.0`. In Spark version 2.4 and earlier,  there is no type checking for it, thus, all type values with a `+` prefix are valid, e.g. `+ array(1, 2)` is valid and results `[1, 2]`. Besides, there is no implicit type coercion for `+` with string, e.g. in Spark 2.4, the result of `+'1'` is string `1`.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: nit: the indentation is wrong. It should be 2 spaces.

Comment (cloud-fan) [reply docs/sql-migration-guide.md]: BTW it's worth mention that, in 2.4 the unary plus operator is ignored. There is no type coercion at all, not only for string type.

**Review by cloud-fan — APPROVED** [review]
LGTM except one minor comment

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
the last commit is just updating migration guide and the previous commit passes all tests. I'm merging it to master, thanks!

**Comment by yaooqinn** [comment]
Thanks for merging

**Pull request closed without merging**